### PR TITLE
[ENG-888] Add feature to spread contacts randomly. EMER.

### PR DESCRIPTION
**Pull request by heathdutton**
For both over-budget and off-schedule contact situations.
This enforces a random distribution of 1~7 days, which is a shift, but
a happy one.

Haven’t tested this at all yet.

### CODE REVIEW

File: Integration/ClientIntegration.php [hunk 1]
```diff
@@ -718,96 +721,145 @@ private function handleException(\Exception $exception)
...
-                    $rescheduleDate->add(new \DateInterval($interval));
-                } catch (\Exception $e) {
+            // To randomly disperse API requests we must get all openings within the date range first.
+            $all = 'api' === $this->contactClient->getType();
+
+            // Get all openings if API, otherwise just get the first available.
+            $openings = $this->payloadModel->getScheduleModel()->findOpening($startDay, $endDay, 1, $all, $startTime);
+            if ($openings) {
+                // Select an opening.
+                if ($all) {
```
Comment (dean-martin) [inline Integration/ClientIntegration.php]: Not using the ternary operator here makes it easier to read. Nice. 💯  👍 ☕️ 😎

File: Integration/ClientIntegration.php [hunk 2]
```diff
@@ -718,96 +721,145 @@ private function handleException(\Exception $exception)
...
-            $this->getSession()->set('contact.client.reschedule.event', $contactClientRescheduleEvents);
+                // Add leadEventLog id instance to global session array for later processing in reschedule() dispatch.
+                $eventLogId          = $this->getEvent()['leadEventLog']->getId();
+                $events              = $this->getSession()->get('contact.client.reschedule.event', []);
+                $events[$eventLogId] = [
+                    'triggerDate' => $start,
+                    'reason'      => $reason,
+                ];
+                $this->getSession()->set('contact.client.reschedule.event', $events);
+                $result = true;
```
Comment (dean-martin) [inline Integration/ClientIntegration.php]: `$result` is redundant here, you can just return true instead.

File: Integration/ClientIntegration.php [hunk 3]
```diff
@@ -718,96 +721,145 @@ private function handleException(\Exception $exception)
...
+            }
         }
+
+        return $result;
```
Comment (dean-martin) [inline Integration/ClientIntegration.php]: This will always be false, so you can just return false instead.

File: Model/FilePayload.php [hunk 4]
```diff
@@ -748,25 +748,27 @@ private function addContactToQueue()
...
                 ->setTimezone();
 
-            list($start, $end) = $this->scheduleModel->nextOpening($rate, $seekDays);
-
-            if (!$start) {
+            $openings = $this->scheduleModel->findOpening(0, $endDay, $rate);
+            if (!$openings) {
                 throw new ContactClientException(
-                    'Could not find an open time slot to send in the next '.$seekDays.' days',
+                    'Could not find an open time slot to send in the next '.$endDay.' days',
```
Comment (dean-martin) [inline Model/FilePayload.php]: I'd consider replacing this with double quotes just for readability with $endDay

File: Model/Schedule.php [hunk 5]
```diff
@@ -143,33 +150,36 @@ public function nextOpening($fileRate, $seekDays, $startDay = 0)
...
-                if ($fileCount > 0 && $fileRate > 1) {
-                    // Push the start time to the next available slot in this day.
-                    $daySeconds = $end->format('U') - $start->format('U');
-                    if ('00:00' === $timeFrom && '23:59' === $timeTill) {
-                        // Avoid sending 2 files at midnight.
-                        $segmentSeconds = intval($daySeconds / $fileRate);
-                    } else {
-                        // Send at opening and closing times, spreading the rest of the day evenly.
-                        $segmentSeconds = intval($daySeconds / ($fileRate - 1));
+                if ('file' === $this->contactClient->getType()) {
```
Comment (dean-martin) [inline Model/Schedule.php]: Yoda comparisons 👌 🔥

File: Model/Schedule.php [hunk 6]
```diff
@@ -143,33 +150,36 @@ public function nextOpening($fileRate, $seekDays, $startDay = 0)
...
 
-                // Return the next appropriate window to send contacts.
-                return [$start, $end];
-                break;
+                $openings[] = [$start, $end];
+                if (!$all) {
+                    break;
+                }
             } catch (\Exception $e) {
                 if ($e instanceof ContactClientException) {
```
Comment (dean-martin) [inline Model/Schedule.php]: You can have multiple catches for a try catch.
`try { // .... }

catch (\Exception $e) { // ... }

catch (ContactClientException) { // ...  }`

File: Model/Schedule.php [hunk 7]
```diff
@@ -179,29 +189,32 @@ public function nextOpening($fileRate, $seekDays, $startDay = 0)
...
-            $day = intval($this->now->format('N')) - 1;
+            $day = intval($date->format('N')) - 1;
             if (isset($hours[$day])) {
                 if (
                     isset($hours[$day]->isActive)
                     && !$hours[$day]->isActive
                 ) {
                     throw new ContactClientException(
-                        'This contact client does not allow contacts on a '.$this->now->format('l').'.',
+                        'This contact client does not allow contacts on a '.$date->format('l').'.',
```
Comment (dean-martin) [inline Model/Schedule.php]: Double quotes would make it easier to read imo.

File: Views/ContactClient/datatable.html.php [hunk 8]
```diff
@@ -20,6 +20,6 @@
     var tableData = <?php echo json_encode($tableData); ?>;
 </script>
 <?php
-    echo $view['assets']->includeScript('plugins/MauticContactClientBundle/Assets/build/contactclient.min.js?v=2', 'contactclientOnLoad', 'contactclientOnLoad');
+    echo $view['assets']->includeScript('plugins/MauticContactClientBundle/Assets/build/contactclient.min.js?v=2.15.1', 'contactclientOnLoad', 'contactclientOnLoad');
```
Comment (dean-martin) [inline Views/ContactClient/datatable.html.php]: Is it possible to pull the version from a config so this doesn't have to be updated ?

File: Views/ContactClient/form.html.php [hunk 9]
```diff
@@ -21,7 +21,11 @@
     $view['translator']->trans('mautic.contactclient.new');
 $view['slots']->set('headerTitle', $header);
 
-echo $view['assets']->includeScript('plugins/MauticContactClientBundle/Assets/build/contactclient.min.js?v=2', 'contactclientOnLoad', 'contactclientOnLoad');
+echo $view['assets']->includeScript(
+    'plugins/MauticContactClientBundle/Assets/build/contactclient.min.js?v=2.15.1',
```
Comment (dean-martin) [inline Views/ContactClient/form.html.php]: Same as above.

**Comment by heathdutton** [comment]
> So without a default of 3, for citizens, it will take leads after 6pm on fri and they'll queue to monday? unless they change the client setting to a greater number?

This is why I made the default at 3, because the most common use will be a closed weekend.
We probably should have a javascript warning of some kind to indicate the possibility of a gap, but I think there's always the possibility of a gap, when taking exclusions and unpublish dates into account.

File: EventListener/ContactClientSubscriber.php [hunk 10]
```diff
@@ -241,27 +243,56 @@ public function onEventScheduled(ScheduledEvent $event)
...
-                )) {
-                // get leadEventLog repo and save log entity.
-                $log->setTriggerDate($contactClientRescheduleEvents[$log->getId()]);
-                $log->setIsScheduled(true);
-                /** @var LeadEventLogRepository $leadEventLogRepo */
-                $leadEventLogRepo = $this->em->getRepository('MauticCampaignBundle:LeadEventLog');
-                $leadEventLogRepo->saveEntity($log);
+            if (!empty($events) && isset($events[$logId])) {
+                if ($events[$logId] instanceof \DateTime) {
+                    // For BC.
```
[the remark this answers is not in the document]
Comment (heathdutton) [reply EventListener/ContactClientSubscriber.php]: Backward compatibility. For a short time in some tasks the shape of the array will be different after deployment.

File: Views/ContactClient/form.html.php (hunk 9 above)
Comment (heathdutton) [reply Views/ContactClient/form.html.php]: Agreed!

**Pull request merged**